### ISSUE TITLE
How does a module provide and force use of a layout?

### ISSUE THREAD

**Issue by docwilmot** [backdrop/backdrop-issues#1045]
Previously, in `hook_preprocess_page` you could add `theme_hook_suggestions` to completely take over the "layout" template of a page. Can you do same in `hook_preprocess_layout`? 

I have a specific path which needs to use a specific page template but it cannot be set with the layouts UI. Can I programmatically create a Layout with that specific template and force it to be used on that path? Or can one skip Layout modules rendering completely and use a hard-coded template?

**Comment by quicksketch** [comment]
There a few approaches you could take:

- You could ship your module with a layout.layout.[name].json file in your module's default config, with the blocks you wanted laid out an assigned the layout you desire. If this module provides a dashboard of some kind or a front-end page, this approach may be better.
- If this is an administrative page with a simple layout like a 2-col arrangement, you may be better off using the module's CSS to provide the layout within the default single-column template used on admin pages.

I've been thinking about this same problem and whether we'd want to use layouts for something like the node add/edit form to make a 2-col layout. I also want to think about how we would handle a layout within a dialog, but in that case, I'm almost certain that we would hard-code the layout within dialog.

**Comment by docwilmot** [comment]
The module is a port of [setup](https://www.drupal.org/project/setup) module, which provides a wizard for setting up site variables and other content after install, so the template content would only be really applicable to the site setup process, so I wouldnt want the layout to be visible in the UI either. (a 'hidden' flag for layouts would have been useful!)

So I've used the technique of the maintenance page and printed the output of the setup page callback directly:
```
print theme('setup', 'stuff' => $stuff);
exit();
```

And then set a template file in `hook_theme`. So far it works. Don't quite know whether this will cause issues yet though.

**Comment by herbdool** [comment]
I can figure out how to include a default layout, but for the life of me I can't figure out how to get the module to use that layout for given contexts/paths. I've been scouring the api for awhile and never seem to get exactly what I'm looking for. Just like going to admin pages will force the use of the admin_default layout, I'd like going to civicrm admin pages go to a new default layout I've added in the module's default config.

@quicksketch @docwilmot are you able to point me in the right direction?

**Comment by docwilmot** [comment]
@herbdool you would make a default layout (see `layout.layout.default.json` for example) and set the path. If all civi paths are similar (like `civicrm/%`) then all those paths would use that layout. Otherwise create layouts for each unique (set of) paths.

**Comment by herbdool** [comment]
Thank you @docwilmot! I knew I was missing something simple to get me part of the way there. For some reason civicrm/% wasn't working but civicrm on it's own is working for all paths below it too.

The other issue is that I can't create layouts for each unique set of paths since it's too varied. I would like to have it check the contexts much like the civicrm submodule checks if it should show the admin theme or not. So with this logic being true: `CRM_Utils_Array::value('is_public', $item)` it should show the public civicrm theme and layout (usually same as the public facing one) and if false show the civicrm admin theme and layout.

It's more than just access control BTW since a "public page" should show the same layout and theme regardless of the user's role.

Let me know if I'm highjacking your issue. I can create a new one. Thank you again.

**Comment by docwilmot** [comment]
In that case you'd probably need two Civi layouts (public/private) and a custom access plugin which evaluates `CRM_Utils_Array::value('is_public', $item)`. See `front_layout_access.inc` for example.

**Comment by herbdool** [comment]
I've made progress with your help @docwilmot. I've got an access plugin and wondering if I should have a context plugin too to ensure the condition is only available on civicrm paths.

In order to get it working properly I had to use `hook_layout_api()` to load a `civicrmtheme.layout.inc` file. And in that file use `hook_layout_access_info()` to register the plugin and also allow someone to revert the layout with `hook_layout_revert()`.

Looks like it's possible to create multiple layouts with the same path, so I may need to ensure they play nicely together. Which makes me wonder how the default admin layout has no setting for path at all and I should adopt the same for CiviCRM.
